File: packages/build-spec/test/context_test.exs
defmodule FrameshiftBuild.ContextTest do
  @moduledoc false
  use ExUnit.Case, async: true

  @root Path.join(__DIR__, "fixtures/context-v1")
  @assembly File.read!(Path.join(@root, "assembly.json"))
  @profiles Enum.map(0..3, &File.read!(Path.join(@root, "profile-#{&1}.json")))
  @mappings Enum.map(0..1, &File.read!(Path.join(@root, "mapping-#{&1}.json")))
  @canonical File.read!(Path.join(@root, "context.json"))
  @identity "sha256:6c3d7e88b4bdb95515f95d28448d3814e1b798c6d7203db3c825c28d17451e23"

  test "exact scope and source bindings produce the browser golden context" do
    assert {:ok, value} = FrameshiftBuild.resolve_context(@assembly, @profiles, @mappings)
    assert value.identity == @identity
    assert value.canonical == @canonical
    assert {:ok, value.assembly_identity} == FrameshiftBuild.build_identity(@assembly)
    assert {:context, {:resolution, _, [_, _, _, _], []}, [_, _]} = value.context

    assert {:ok, ^value} =
             FrameshiftBuild.resolve_context(
               @assembly,
               Enum.reverse(@profiles),
               Enum.reverse(@mappings)
             )
  end

  test "mapping sources change context without rewriting assembly identity" do
    [first, second] = @mappings
    changed = String.replace(first, "test-1", "test-2")
    assert {:ok, value} = FrameshiftBuild.resolve_context(@assembly, @profiles, [changed, second])
    refute value.identity == @identity
    assert {:ok, value.assembly_identity} == FrameshiftBuild.build_identity(@assembly)
  end

  test "stale scope, ambiguous revisions and absent ports refuse" do
    [first, _] = @mappings
    assert {:ok, scope} = FrameshiftBuild.inspect_mapping(first)

    for changed <- [
          String.replace(first, scope.profile, "sha256:" <> String.duplicate("9", 64)),
          String.replace(first, "\"firmware\":\"#{scope.firmware}\"", "\"firmware\":\"other\""),
          String.replace(first, "\"input\":\"in\"", "\"input\":\"absent\"")
        ] do
      assert {:error, "invalid_reference"} =
               FrameshiftBuild.resolve_context(@assembly, @profiles, [changed])
    end

    assert {:error, "duplicate_identifier"} =
             FrameshiftBuild.resolve_context(@assembly, @profiles, [first, first])

    assert {:error, "duplicate_identifier"} =
             FrameshiftBuild.resolve_context(@assembly, @profiles, [
               first,
               String.replace(first, "test-1", "test-2")
             ])

    assert {:error, "noncanonical"} =
             FrameshiftBuild.resolve_context(@assembly, @profiles, [first <> " "])
  end

  test "unresolved profiles remain explicit and cannot support supplied mappings" do
    assert {:ok, value} = FrameshiftBuild.resolve_context(@assembly, [], [])
    assert {:context, {:resolution, _, [], [_, _, _, _]}, []} = value.context
    refute value.identity == @identity

    assert {:error, "invalid_reference"} =
             FrameshiftBuild.resolve_context(@assembly, [], @mappings)
  end

  test "combined document budgets and input types fail before parsing" do
    for bad <- [nil, %{}, [nil], ["body" | :improper], [%{identity: "fake", bytes: "{}"}]] do
      assert {:error, "invalid_document"} =
               FrameshiftBuild.resolve_context(@assembly, @profiles, bad)

      assert {:error, "invalid_document"} =
               FrameshiftBuild.resolve_context(@assembly, bad, @mappings)
    end

    assert {:error, "invalid_document"} = FrameshiftBuild.resolve_context(nil, [], [])

    assert {:error, "invalid_count"} =
             FrameshiftBuild.resolve_context(@assembly, [], List.duplicate("", 65))

    assert {:error, "too_large"} =
             FrameshiftBuild.resolve_context(@assembly, [], [String.duplicate("x", 262_145)])

    bodies = List.duplicate(String.duplicate("x", 262_144), 8)
    assert {:error, "too_large"} = FrameshiftBuild.resolve_context(@assembly, bodies, bodies)
  end
end

File: packages/build-spec/test/context.test.mjs
import assert from 'node:assert/strict';
import {readFile} from 'node:fs/promises';
import {test} from 'node:test';
import {resolveContext} from '../js/context.mjs';

const read=name=>readFile(new URL(`./fixtures/context-v1/${name}.json`,import.meta.url),'utf8');
const assembly=await read('assembly');
const profiles=await Promise.all([0,1,2,3].map(n=>read('profile-'+n)));
const mappings=await Promise.all([0,1].map(n=>read('mapping-'+n)));
const canonical=await read('context');
const identity='sha256:6c3d7e88b4bdb95515f95d28448d3814e1b798c6d7203db3c825c28d17451e23';

test('verified context has the same canonical bytes and identity as BEAM',async()=>{
  const result=await resolveContext(assembly,profiles,mappings);
  assert.equal(result.ok,true);
  assert.equal(result.identity,identity);
  assert.equal(result.canonical,canonical);
  assert.equal(result.assembly_identity,JSON.parse(canonical).assembly);
  assert.deepEqual(result,await resolveContext(assembly,[...profiles].reverse(),[...mappings].reverse()));
  assert.equal(result.context.mappings.toArray().length,2);
  assert.deepEqual(result.context.resolution.missing.toArray(),[]);
  const changed=await resolveContext(assembly,profiles,[mappings[0].replace('test-1','test-2'),mappings[1]]);
  assert.equal(changed.ok,true);
  assert.notEqual(changed.identity,identity);
  assert.equal(changed.assembly_identity,result.assembly_identity);
});

test('stale scope, duplicate revisions, nonexistent ports and changed profiles refuse',async()=>{
  const changed=JSON.parse(mappings[0]);
  for(const document of [
    {...changed,firmware:'other'},
    {...changed,profile:'sha256:'+'9'.repeat(64)},
    {...changed,pairs:[{input:'absent',output:'out'}]},
  ]) assert.deepEqual(await resolveContext(assembly,profiles,[JSON.stringify(document)+'\n']),{ok:false,error:'invalid_reference'});
  assert.deepEqual(await resolveContext(assembly,profiles,[mappings[0],mappings[0]]),{ok:false,error:'duplicate_identifier'});
  assert.deepEqual(await resolveContext(assembly,profiles,[mappings[0],mappings[0].replace('test-1','test-2')]),{ok:false,error:'duplicate_identifier'});
  assert.deepEqual(await resolveContext(assembly,profiles,[mappings[0]+' ']),{ok:false,error:'noncanonical'});
  assert.deepEqual(await resolveContext(assembly,[profiles[0].replace('"part_revision":"R1"','"part_revision":"R2"'),...profiles.slice(1)],mappings),{ok:false,error:'unreferenced_profile'});
});

test('missing bodies stay explicit while supplied maps cannot resolve missing profiles',async()=>{
  const result=await resolveContext(assembly,[],[]);
  assert.equal(result.ok,true);
  assert.equal(result.context.resolution.missing.toArray().length,4);
  assert.equal(result.context.mappings.toArray().length,0);
  assert.notEqual(result.identity,identity);
  assert.deepEqual(await resolveContext(assembly,[],mappings),{ok:false,error:'invalid_reference'});
});

test('combined budgets and untrusted collection types fail before cryptography',async context=>{
  let hashes=0;
  context.mock.method(globalThis.crypto.subtle,'digest',async()=>{hashes++;throw new Error('unexpected');});
  for(const bad of [null,{},[undefined],new Array(1),[{identity:'fake',bytes:mappings[0]}]]) {
    assert.deepEqual(await resolveContext(assembly,profiles,bad),{ok:false,error:'invalid_document'});
    assert.deepEqual(await resolveContext(assembly,bad,mappings),{ok:false,error:'invalid_document'});
  }
  assert.deepEqual(await resolveContext(null,[],[]),{ok:false,error:'invalid_document'});
  assert.deepEqual(await resolveContext(assembly,[],Array(65).fill('')),{ok:false,error:'invalid_count'});
  assert.deepEqual(await resolveContext(assembly,Array(65).fill(''),[]),{ok:false,error:'invalid_count'});
  assert.deepEqual(await resolveContext(assembly,[],['x'.repeat(262145)]),{ok:false,error:'too_large'});
  assert.deepEqual(await resolveContext(assembly,[],['é'.repeat(131073)]),{ok:false,error:'too_large'});
  assert.deepEqual(await resolveContext(assembly,Array(8).fill('x'.repeat(262144)),Array(8).fill('x'.repeat(262144))),{ok:false,error:'too_large'});
  assert.equal(hashes,0);
});

test('async hashing uses a stable snapshot of both caller arrays',async context=>{
  const mutableProfiles=[...profiles];
  const mutableMappings=[...mappings];
  const digest=globalThis.crypto.subtle.digest.bind(globalThis.crypto.subtle);
  context.mock.method(globalThis.crypto.subtle,'digest',async(...args)=>{
    mutableProfiles.splice(0,mutableProfiles.length,'x'.repeat(262145));
    mutableMappings.splice(0,mutableMappings.length,'{}');
    return digest(...args);
  });
  const result=await resolveContext(assembly,mutableProfiles,mutableMappings);
  assert.equal(result.ok,true);
  assert.equal(result.identity,identity);
});

test('cryptography loss refuses even an unresolved context',async context=>{
  context.mock.method(globalThis.crypto.subtle,'digest',async()=>{throw new Error('offline');});
  assert.deepEqual(await resolveContext(assembly,[],[]),{ok:false,error:'crypto_unavailable'});
});
